src/datatrove/pipeline/inference/run_inference.py

```diff
@@ -637,78 +624,32 @@ async def _handle_record(
-                    # It's an async generator - process each payload as soon as it's yielded
-                    async for payload in payloads_result:
-                        if payload is None:
-                            continue
-                        callback_requested = callback_requested or payload.pop("callback", False) is True
-                        # Set default values for payload
-                        payload.setdefault("model", self.config.model_name_or_path)
-                        payload.setdefault("temperature", self.config.temperature)
-
-                        # Start request immediately
-                        task = asyncio.create_task(self._send_request(payload, semaphore))
-                        request_tasks.append(task)
-                    if not request_tasks:
-                        return
-
-                elif isinstance(payloads_result, dict):
-                    # Single dict
-                    payload = payloads_result
-                    callback_requested = callback_requested or payload.pop("callback", False) is True
-                    payload.setdefault("model", self.config.model_name_or_path)
-                    payload.setdefault("temperature", self.config.temperature)
-                    task = asyncio.create_task(self._send_request(payload, semaphore))
-                    request_tasks.append(task)
-                elif payloads_result is None:
-                    return
-
-                if not request_tasks:
-                    raise InferenceProcessingError(doc, "No valid payloads generated from query_builder")
-
-                # Wait for all requests to complete and collect results in order
-                results = await asyncio.gather(*request_tasks)
-
-                for result in results:
-                    if isinstance(result, InferenceError) and (
-                        not self.skip_bad_requests or "BadRequestError" not in result.error
-                    ):
-                        # re-raise any non-skippable errors
-                        raise InferenceProcessingError(doc, result.error)
-
-                # Store results directly in document metadata
-                doc.metadata.setdefault("inference_results", []).extend(results)
-
-                # Post-process the document if a function is provided. We still want the actual document for checkpointing purposes.
-                if self.postprocess_fn:
-                    postprocess_result = self.postprocess_fn(self, doc)
-                    if postprocess_result is None:
-                        doc.metadata["postprocess_remove"] = True
-                    else:
-                        doc = postprocess_result
-
-                if callback_requested:
-                    await _handle_record(doc, rank, chunk_index, output_writer_context)
-                    return
+                generate_callback = partial(self._send_request, semaphore=semaphore)
```

Comment:
No need to  be in try block and overall this object needs to be created just once when we are initializing the semaphore. Move it out of the fc and then just pass the generate_callback instead of the semaphore

Reply:
> Move it out of the fc and then just pass the generate_callback instead of the semaphore

you're right if it stays as is, but for the cache I will make this function receive doc_id and rollout_idx, so I will need to call partial each time anyway

> I don't like the fact that the InferenceError raise from any rollout cause all rollouts to fail. the try/except should be applied per rollout not on all of them. Also is that intended that any inference error causes whole thing to die ?

the idea is that the user should be the one handling generation errors now since they call the function. If they don't, the whole thing dies